mistralrs-core/src/gguf/gguf_tokenizer.rs

```diff
@@ -140,15 +183,43 @@ fn unigram_tokenizer(p: &PropsGGUF) -> Result<(Tokenizer, TokenizerKind, Vec<Str
         .build()?;
 
     // Add special tokens (bos, eos, unk):
-    let mut special_tokens = Vec::<String>::new();
-    for token_id in [bos, eos, unk] {
-        let token = p.tokens[token_id as usize].as_str();
+    let special_tokens = add_special_tokens(p, &mut tokenizer, bos, eos, Some(unk));
 
-        special_tokens.push(token.to_owned());
-        tokenizer.add_special_tokens(&[AddedToken::from(token.to_owned(), true)]);
+    Ok((tokenizer, TokenizerKind::Unigram, special_tokens))
+}
+
+fn bpe_tokenizer(p: &PropsGGUF) -> Result<(Tokenizer, TokenizerKind, AddedTokensCollection)> {
+    let merges = p
+        .merges
+        .as_ref()
+        .ok_or(anyhow::Error::msg("BPE tokenizer must include merges"))?
+        .iter()
+        .map(|merge| {
+            let split: (&str, &str) = merge
+                .splitn(2, ' ')
+                .collect_tuple()
+                .expect("Failed to convert split into 2-tuple");
+            (split.0.to_string(), split.1.to_string())
+        })
+        .collect::<Vec<_>>();
+
+    let mut vocab = HashMap::new();
+    for (i, token) in p.tokens.iter().enumerate() {
+        #[allow(clippy::cast_possible_truncation)]
+        vocab.insert(token.clone(), i as u32);
     }
 
-    Ok((tokenizer, TokenizerKind::Unigram, special_tokens))
+    let bpe = BpeBuilder::new()
+        .vocab_and_merges(vocab, merges)
+        .build()
+        .map_err(anyhow::Error::msg)?;
+    let mut tokenizer = Tokenizer::new(ModelWrapper::BPE(bpe));
+    tokenizer.with_decoder(decoders::byte_level::ByteLevel::new(true, true, true));
```

Comment:
While not necessary, this could be handled with the `TokenizerX::builder()`, which you could then call `.with_model(bpe)`.

The `ByteLevel` decoder would need to be added to the decoder enum below with:

```rs
enum Decoder<'a> {
    ByteLevel(bool, bool, bool),
```

```rs
    fn try_from(variant: Decoder) -> Result<Self, Self::Error> {
        let value: DecoderWrapper = match variant {
            Decoder::ByteLevel(add_prefix_space, trim_offsets, use_regex) => {
                ByteLevel::new(add_prefix_space, trim_offsets, use_regex).into()
            }
```

Then you would have:

```suggestion
    let mut tokenizer = TokenizerX::builder()
        .with_model(bpe);
        .with_decoder(Decoder::ByteLevel(true, true, true));
```

That would make it a bit more aligned with the unigram construction.

Reply:
Thanks for the idea, I have added this.